src/sources/syslog.rs

```diff
@@ -219,59 +224,110 @@ impl SyslogDecoder {
-                .map_err(|_| {
-                    LinesCodecError::Io(io::Error::new(
-                        io::ErrorKind::InvalidData,
-                        "Unable to decode message len as number",
-                    ))
-                })?;
-
-            let from = i + 1;
-            let to = from + len;
-
-            if let Some(msg) = src.get(from..to) {
-                let s = std::str::from_utf8(msg)
-                    .map_err(|_| {
-                        LinesCodecError::Io(io::Error::new(
-                            io::ErrorKind::InvalidData,
-                            "Unable to decode message as UTF8",
-                        ))
-                    })?
-                    .to_string();
-                src.advance(to);
-                Ok(Some(s))
-            } else {
-                Ok(None)
+        loop {
+            let read_to = cmp::min(self.other.max_length().saturating_add(1), src.len());
+            let space_pos = src.iter().position(|&b| b == b' ');
+
+            // If we are discarding, discard to the next newline.
+            let newline_pos = src.iter().position(|&b| b == b'\n');
+
+            match (self.is_discarding, newline_pos, space_pos) {
+                (true, Some(offset), _) => {
+                    // When discarding we keep discarding to the next newline.
+                    src.advance(offset + 1);
+                    self.is_discarding = false;
+                    self.in_octet = false;
+                    return Err(LinesCodecError::Io(io::Error::new(
+                        io::ErrorKind::Other,
+                        "Frame length limit exceeded",
+                    )));
+                }
+                (true, None, _) => {
+                    // There is no newline in this frame. Advance as far as we can to
+                    // discard the entire frame.
+                    src.advance(read_to);
+                    if src.is_empty() {
+                        return Ok(None);
+                    }
+                }
+                (false, _, Some(i)) => {
+                    let len: usize = match std::str::from_utf8(&src[..i])
+                        .map_err(|_| ())
+                        .and_then(|num| num.parse().map_err(|_| ()))
+                    {
+                        Ok(len) => len,
+                        Err(_) => {
+                            // Advance the buffer past the erroneous bytes
+                            // to prevent us getting stuck in an infinite loop.
+                            src.advance(i + 1);
+                            self.in_octet = false;
+                            return Err(LinesCodecError::Io(io::Error::new(
+                                io::ErrorKind::InvalidData,
+                                "Unable to decode message len as number",
+                            )));
+                        }
+                    };
+
+                    let from = i + 1;
```

Comment:
I'd rename `i` to `space_pos`, otherwise the context is quickly lost in this relatively large branch.